### ISSUE TITLE
obs_frontend_set_current_scene crashes if not used in Studio Mode

### ISSUE THREAD

**Issue by woganmay** [obsproject/obs-studio#3385]
## Platform
Operating system and version: Windows 10 v1909 Build 18363.1016
OBS Studio version: 25.0.8 (64 bit)

## Expected Behavior
I'm using a Lua script to automatically change a scene after a certain period of time - see link here: https://gist.github.com/woganmay/78bad7577b52d0fbf09b2cefcd8a131f

I'm expecting that, after switching to Scene A, that OBS will automatically switch to Scene B after 5000 milliseconds. The callbacks and countdown timer work normally, and the scene does transition if I'm in Studio Mode.

## Current Behavior
If I trigger this script while not in Studio mode, OBS will crash when attempting to call obs_frontend_set_current_scene. If I'm in Studio Mode, the scene switches normally.

## Steps to Reproduce
1. Load the Lua plugin linked above
2. Create Scene A, Scene B and a third scene
3. Without Studio Mode enabled, switch to Scene A and wait 3 seconds
4. OBS becomes unresponsive and has to be force-quit
5. Re-launch OBS, enable Studio Mode, stage Scene A and Transition to it
6. Within 3 seconds, the current scene switches to Scene B

Video: https://www.youtube.com/watch?v=ljFxT26Pi9g

## Additional information
None

**Comment by Fenrirthviti** [comment]
Can you also include the crashlog?

**Comment by woganmay** [comment]
There aren't any crash logs. The last crash I have in my folder is from June:

![image](https://user-images.githubusercontent.com/225550/91908140-62d65d80-ecab-11ea-96fa-796b26139183.png)

Is there another place I can get that from?

**Comment by Scrxtchy** [comment]
Doesn't crash for me, OBS stops responding

**Comment by Scrxtchy** [comment]
It appears that in the OBS stops responding between within this block of code
https://github.com/obsproject/obs-studio/blob/master/UI/window-basic-main-transitions.cpp#L701-L704
In doing so, it's sending an event for `OBS_FRONTEND_EVENT_PREVIEW_SCENE_CHANGED`, where this should be sending a `OBS_FRONTEND_EVENT_SCENE_CHANGED`. It is likely that due to this, and the program being not being in PreviewProgram mode, something is bricking the UI

**Comment by WizardCM** [comment]
@Scrxtchy Hmm, my wild guess is that highlighted block should have a `IsPreviewProgramMode()` check on it too (similar to further up in the function).

**Comment by Scrxtchy** [comment]
It's odd, because the automaticsceneswitcher actually also passes over this block, but has no ill effect on OBS

**Comment by Scrxtchy** [comment]
I think this is actually working as intended, the new rabbithole is that it crashes in the obs-scripting callback

**Comment by Scrxtchy** [comment]
Honestly I don't know where to go from there, 
It's only in this exact case where this happens, the function that's called is the lock_callback() on https://github.com/obsproject/obs-studio/blob/81e4211f64c7c2e8c8d1043f63fa1df2cf16dc89/deps/obs-scripting/obs-scripting-lua-frontend.c#L198, before finally giving up on Line 127 of pthread_mutex_lock.c, one of the dependencies

**Comment by Scrxtchy** [comment]
I can indeed also confirm this is an issue with the python scripting, I've optimised the provided script to be able to test it easier. You can pick the two sources instead of relying on them to be hardcoded
https://gist.github.com/Scrxtchy/ee86525b3286272e1f57e92e0c82e8e5

I am very tired now

**Comment by Scrxtchy** [comment]
@woganmay if you're still tackling a mitigation for this issue, I have managed to sidestep the issue on the scripting side.
As I expect this to be an issue of locking the call-back that is currently active, a simple solution would be to remove the frontend callback to switch scenes, and recreating it when the process is complete.

This would make your `set_current_scene` function expanded to look like this
```
obs.obs_frontend_remove_event_callback(on_event)
obs.obs_frontend_set_current_scene(targetScene)
obs.obs_frontend_add_event_callback(on_event)
```
In short, we can't have recursion of call-backs in scripting, it locks the thread

**Comment by woganmay** [comment]
Ahh, that makes sense! Thanks for looking into this, I appreciate it.

For now we're mitigating by just never not being in studio mode, but I'll give that workaround a try.

**Comment by RytoEX** [comment]
I can sort of confirm this. I do not experience a crash, but OBS Studio 28.0.3 does hang after following the STR in the initial post.

**Comment by PatTheMav** [comment]
This is yet another issue caused by script activity being triggered by the graphics thread as well as the main thread:

* Timers are executed by the `lua_tick` function, which is called on every tick of the graphics thread
* Executing the timer callback function requires the callback lock
* With this lock set, `autoswitch_timer_callback` is executed
* The callback itself calls `obs_frontend_set_current_scene`, which schedules a call to `SetCurrentScene` on the main thread
* `SetCurrentScene` in turn signals the `OBS_FRONTEND_EVENT_PREVIEW_SCENE_CHANGED` event
* Our scripting plugin registers itself as a handler for frontend events, so the event is handled by `frontend_event_callback` _on the main thread_
* This function now tries to acquire callback lock, which makes it wait, because the graphics thread already holds the lock

So the graphics thread executes the timer callback, which waits for the scene change to finish, but the scene change has to wait for the lock held by the graphics thread to be released. A classic deadlock.

We could investigate if scene changes triggered by scripts need to be blocking (or could just be fire&forget), which could fix this issue.
